flocker/node/script.py

```diff
@@ -281,143 +319,282 @@ def get_service(self, reactor, options):
+    :return: The ``VolumeService``, started.
     """
-    Given a dataset configuration, return a dataset deployer factory.
+    pool = zfs.StoragePool(
+        reactor=reactor, name=name, mount_root=FilePath(mount_root),
+    )
+    api = VolumeService(
+        config_path=FilePath(volume_config_path),
+        pool=pool,
+        reactor=reactor,
+    )
+    api.startService()
+    return api
+
+
+class DeployerType(Names):
+    """
+    References to the different ``IDeployer`` implementations that are
+    available.
+
+    :ivar p2p: The "peer-to-peer" deployer - suitable for use with systems like
+        ZFS where nodes interact directly with each other for data movement.
+    :ivar block: The Infrastructure-as-a-Service deployer - suitable for use
+        with system like EBS where volumes can be attached to nodes as block
+        devices and then detached (and then re-attached to other nodes).
+    """
+    p2p = NamedConstant()
+    block = NamedConstant()
+
+
+class BackendDescription(PRecord):
+    """
+    Represent one kind of storage backend we might be able to use.
+
+    :ivar name: The human-meaningful name of this storage backend.
+    :ivar needs_reactor: A flag which indicates whether this backend's API
+        factory needs to have a reactor passed to it.
+    :ivar needs_id: A flag which indicates whether this backend's API factory
+        needs to have the cluster's unique identifier passed to it.
+    :ivar api_factory: An object which can be called with some simple
+        configuration data and which returns the API object implementing this
+        storage backend.
+    :ivar deployer_type: A constant from ``DeployerType`` indicating which kind
+        of ``IDeployer`` the API object returned by ``api_factory`` is usable
+        with.
+    """
+    name = field(type=unicode, mandatory=True)
+    needs_reactor = field(type=bool, mandatory=True)
+    # XXX Eventually everyone will take cluster_id so we will throw this flag
+    # out.
+    needs_id = field(type=bool, mandatory=True)
+    api_factory = field(mandatory=True)
+    deployer_type = field(mandatory=True)
+
+# These structures should be created dynamically to handle plug-ins
+_DEFAULT_BACKENDS = [
+    # P2PManifestationDeployer doesn't current know anything about
+    # cluster_uuid.  It probably should so that it can make sure it
+    # only talks to other nodes in the same cluster (maybe the
+    # authentication layer would mostly handle this but maybe not if
+    # you're slightly careless with credentials - also ZFS backend
+    # doesn't use TLS yet).
+    BackendDescription(
+        name=u"zfs", needs_reactor=True, needs_id=False,
+        api_factory=_zfs_storagepool, deployer_type=u"p2p",
+    ),
+    BackendDescription(
+        name=u"loopback", needs_reactor=False, needs_id=False,
+        api_factory=LoopbackBlockDeviceAPI.from_path, deployer_type=u"block",
```

Comment:
The docs for BackendDescription say that `deployer_type` should be one of the `DeployerType` `NamedConstant` s.